genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/S3ArchivalServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+@Component
+@Lazy
+@Slf4j
+class S3ArchivalServiceImpl implements ArchivalService {
+
+    private final AmazonS3 amazonS3;
+
+    S3ArchivalServiceImpl(final AmazonS3 amazonS3) {
+        this.amazonS3 = amazonS3;
+    }
+
+    /**
+     * Store the contents of a file to S3.
+     * For a directory all its contents are recursively traversed. Each of the file is
+     * stored as a an object in S3. Each empty directory is also stored as an object with
+     * empty content.
+     * <p>
+     * Example 1
+     * <p>
+     * For a dir hierarchy
+     * /tmp/${jobid}/genie/applications (empty dir)
+     * /tmp/${jobid}/genie/run (file)
+     * /tmp/${jobid}/genie/commands/presto (file)
+     * <p>
+     * dir to be archived - /tmp/${jobid}
+     * target location - s3://$bucketName/{$jobid}
+     * <p>
+     * following structure is created in S3
+     * s3://$bucketName/{$jobid}/genie/applications/ -> empty content
+     * s3://$bucketName/{$jobid}/genie/run -> contents of run file
+     * s3://$bucketName/{$jobid}/genie/commands/presto -> contents of presto file
+     * <p>
+     * Example 2
+     * <p>
+     * file to archive - /tmp/run
+     * target location - s3://$bucketName/run
+     * <p>
+     * following structure is created in S3
+     * s3://$bucketName/abc -> contents of run file
+     *
+     * @param file              file/directory to archive.
+     * @param targetLocationURI target location uri for the file
+     * @throws ArchivalException if archival fails
+     */
+
+    @Override
+    public void archive(
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws ArchivalException {
+        log.info("Archiving to location: {}", targetLocationURI);
+
+        try {
+            final AmazonS3URI targetLocationAmazonS3URI = new AmazonS3URI(targetLocationURI);
+            archiveRecursively(file, targetLocationAmazonS3URI);
+        } catch (Exception e) {
+            log.info("Error archiving to target location: {} ", targetLocationURI);
+            throw new ArchivalException("Error archiving file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Traverse a directory recursively and archive the files and directories
+     * underneath it. If the directory is empty, create an empty object in S3.
+     * For a regular child file create an object in S3 and copy the file contents.
+     * For a empty child directory create an empty object in S3
+     *
+     * @param fileToArchive     file/dir to archive.
+     * @param targetLocationURI target location URI for storing the file in S3.
+     */
+    private void archiveRecursively(
+        @NotNull final File fileToArchive,
+        @NotNull final AmazonS3URI targetLocationURI
+    ) {
+        final File[] children = fileToArchive.listFiles();
+
+        //children == null if fileToArchive is not a directory.
+        //children is a empty array if the directory is empty
+        if (children == null || children.length == 0) {
```

Comment:
This could be simply:
```
if (children == null) {
  uploadFile
} else {
  createEmptyFolder
  for c in children {
    archiveRecursively
  }
}
```
Combined with Tom's comment below, this should shorten this file a bit and make it easier to test.

Reply:
This will not work for the case where the folder is not empty. It will unnecessary create an entry for an empty folder. Hence we needed the children.empty check i.e 

```if (children == null || children.length == 0) ```
Regardless, since we dont plan to save empty folders now this is redundant and will anyways be simplified in the change I will do

Reply:
Fine, if you want to skip empty folder creation then it is
```
if (children == null) {
  uploadFile
} else if (children.isEmpty) {
  createEmptyFolder
} else {
  for c in children {
    archiveRecursively
  }
}
```